src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,344 @@
+                    }
+                  },
+                  RayConfig::instance().gcs_lease_worker_retry_interval_ms());
+  };
+
+  auto client = GetOrCreateNodeClient(node_id, *node);
+  auto status = client->RequestWorkerLease(
+      request, [this, node_id, retry_leasing_worker, actor](
+                   const Status &status, const rpc::RequestWorkerLeaseReply &reply) {
+        if (auto node = gcs_node_manager_.GetNode(node_id)) {
+          if (!status.ok()) {
+            retry_leasing_worker();
+            return;
+          }
+          RAY_LOG(INFO) << "Finished leasing worker from " << node_id << " for actor "
+                        << actor->GetActorID();
+          HandleWorkerLeasedReply(actor, reply);
+        }
+      });
+
+  if (!status.ok()) {
+    retry_leasing_worker();
+  }
+}
+
+void GcsActorScheduler::HandleWorkerLeasedReply(
+    std::shared_ptr<GcsActor> actor, const ray::rpc::RequestWorkerLeaseReply &reply) {
+  // Remove the actor from the leasing queue.
+  RemoveActorInPhaseOfLeasing(actor);
+
+  const auto &retry_at_raylet_address = reply.retry_at_raylet_address();
+  const auto &worker_address = reply.worker_address();
+  if (worker_address.raylet_id().empty()) {
+    RAY_CHECK(!retry_at_raylet_address.raylet_id().empty());
+    auto address = actor->GetAddress();
+    address.set_raylet_id(retry_at_raylet_address.raylet_id());
+    actor->UpdateAddress(address);
+    auto actor_table_data =
+        std::make_shared<rpc::ActorTableData>(actor->GetActorTableData());
+    // The backend storage is reliable in the future, so the status must be ok.
+    RAY_CHECK_OK(redis_gcs_client_->Actors().AsyncUpdate(
+        actor->GetActorID(), actor_table_data, [this, actor](Status status) {
+          RAY_CHECK_OK(status);
+          Schedule(actor);
+        }));
+  } else {
+    std::vector<rpc::ResourceMapEntry> resources;
+    for (auto &resource : reply.resource_mapping()) {
+      resources.emplace_back(resource);
+    }
+    auto leased_worker = CreateLeasedWorker(worker_address, std::move(resources));
+    RAY_LOG(INFO) << "Worker " << leased_worker->GetWorkerID()
+                  << " is responsible for creating actor " << actor->GetActorID();
+    AddWorkerInPhaseOfCreating(leased_worker);
+    CreateActorOnWorker(actor, leased_worker);
+  }
+}
+
+void GcsActorScheduler::CreateActorOnWorker(std::shared_ptr<GcsActor> actor,
+                                            std::shared_ptr<GcsLeasedWorker> worker) {
+  RAY_CHECK(actor && worker);
+  actor->UpdateAddress(worker->GetAddress());
+  RAY_LOG(INFO) << "Start creating actor " << actor->GetActorID() << " on worker "
+                << worker->GetWorkerID() << " at node " << actor->GetNodeID();
+  std::weak_ptr<GcsLeasedWorker> weak_worker(worker);
+  worker->CreateActor(actor->GetCreationTaskSpecification(), [this, actor, weak_worker] {
+    // This lambda should not capture shared_ptr of the worker, otherwise the worker may
```

Comment:
> I don't think I understand this - shouldn't the lambda be cleaned up at some point (e.g., when the actor dies)? Otherwise we have a resource leak anyways. It's a red flag that we have to use a weak_ptr here.

Actually, `weak_ptr` is very useful sometimes, but since both you and @stephanie-wang confused about this, I'll replace it with shared_ptr, and then check if assigned_actor_id is nil to clean up lambda in the event loop at the bottom layer.